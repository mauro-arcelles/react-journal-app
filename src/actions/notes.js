import { addDoc, collection, deleteDoc, doc, updateDoc } from 'firebase/firestore';
import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';

export const startNewNote = () => {
  return async (dispatch, getState) => {

    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    };

    const doc = await addDoc(collection(db, `${uid}/journal/notes/`), newNote);

    dispatch(activeNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));

  };
};


export const activeNote = (id, note) => {
  return {
    type: types.notesActive,
    payload: {
      id,
      ...note
    }
  };
};

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => {
  return {
    type: types.notesLoad,
    payload: notes
  };

};

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
    await updateDoc(noteRef, noteToFirestore);

    dispatch(refreshNote(note.id, noteToFirestore));
    // dispatch(startLoadingNotes(uid));

    Swal.fire('Saved', 'Your note has been saved', 'success');

  };
};

export const refreshNote = (id, note) => {
  return {
    type: types.notesUpdated,
    payload: {
      id,
      note: {
        id,
        ...note
      }
    }
  };
};

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait',
      allowOutsideClick: false,
      didOpen: () => {
        Swal.showLoading();
      }
    });

    const fileUrl = await fileUpload(file);

    activeNote.url = fileUrl;
    dispatch(startSaveNote(activeNote));

    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;

    const noteRef = doc(db, `${uid}/journal/notes/${id}`);
    await deleteDoc(noteRef);

    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => {
  return {
    type: types.notesDelete,
    payload: id
  };
};

export const noteLogout = () => {
  return {
    type: types.notesLogoutCleaning
  };
};

export const addNewNote = (id, note) => {
  return {
    type: types.notesAddNew,
    payload: {
      id,
      ...note
    }
  };
};