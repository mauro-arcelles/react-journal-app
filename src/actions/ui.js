import { types } from '../types/types';

export const setError = (err) => {
  return {
    type: types.uiSetError,
    payload: err
  };
};

export const removeError = () => {
  return {
    type: types.uiRemoveError
  };
};

export const startLoading = () => {
  return {
    type: types.uiSetLoading
  };
};

export const removeLoading = () => {
  return {
    type: types.uiRemoveLoading
  };
};