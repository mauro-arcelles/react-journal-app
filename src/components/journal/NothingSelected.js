
export const NothingSelected = () => {
  return (
    <div className='nothing__main-content'>
      <div>
        Select something
        <hr />
        or create an entry!
      </div>

      <i className='fa-regular fa-star fa-4x mt-5'></i>
    </div>
  );
};
