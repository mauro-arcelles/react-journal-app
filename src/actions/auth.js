import { types } from '../types/types';
import { auth, googleAuthProvider } from '../firebase/firebase-config';
import { createUserWithEmailAndPassword, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from 'firebase/auth';
import { removeLoading, startLoading } from './ui';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// PARA RECORDAR:
// TODOS LAS ACCIONES ASINCRONAS TIENEN QUE SER UN METODO QUE RETORNA UN CALLBACK
// EN DONDE TENDRÁ EL dispatch Y el getState GRACIAS A redux-thunk


export const startLoginEmailPassword = (email, password) => {

  return (dispatch) => {
    dispatch(startLoading());

    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(removeLoading());
      })
      .catch(e => {
        dispatch(removeLoading());
        Swal.fire('Error', 'Username or password are incorrect', 'error');
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {

    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });

        dispatch(login(user.uid, user.displayName));
      })
      .catch(e => {
        Swal.fire('Error', 'Username already exists', 'error');
      });
  };

};


export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});


export const startLogout = () => {
  return async (dispatch) => {

    await signOut(auth);

    dispatch(logout());
    dispatch(noteLogout());

  };
};

export const logout = () => {
  return {
    type: types.logout
  };
};
