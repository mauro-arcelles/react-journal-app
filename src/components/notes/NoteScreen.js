import { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {

  const dispatch = useDispatch();

  const { active: noteActive } = useSelector(state => state.notes);
  const [formValues, handleInputChange, reset] = useForm(noteActive);

  const { body, title } = formValues;

  const activeId = useRef(noteActive.id);

  useEffect(() => {
    if (noteActive.id !== activeId.current) {
      reset(noteActive);
      activeId.current = noteActive.id;
    }
  }, [noteActive, reset]);

  useEffect(() => {
    dispatch(activeNote(formValues.id, { ...formValues }));

  }, [formValues, dispatch]);

  const handleDelete = () => {
    dispatch(startDeleting(noteActive.id));
  };

  return (
    <div className='notes__main-content' >

      <NotesAppBar />

      <div className='notes__content'>

        <input type="text" placeholder='Some awesome title' className='notes__title-input' autoComplete='off' value={title} onChange={handleInputChange} name='title' />
        <textarea placeholder='What happen today?' className='notes__textarea' autoComplete='off' value={body} onChange={handleInputChange} name='body'></textarea>

        {
          (noteActive.url) &&
          (<div className="notes__image">
            <img src={noteActive.url} alt="imagen" width='100%' height='100%' />
          </div>)
        }

      </div>

      <button className='btn btn-danger' onClick={handleDelete}>
        Delete
      </button>

    </div >
  );
};
