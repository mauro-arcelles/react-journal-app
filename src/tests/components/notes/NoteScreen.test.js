import { mount } from 'enzyme/build';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { activeNote } from '../../../actions/notes';
import { Sidebar } from '../../../components/journal/Sidebar';
import { NoteScreen } from '../../../components/notes/NoteScreen';

jest.mock("../../../actions/notes", () => {
  return {
    activeNote: jest.fn(),
  };
});


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    uid: 'abc123',
    name: 'Adrian'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    active: {
      id: 'abc123',
      title: 'Note title',
      body: 'Note body',
    },
    notes: []
  }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <NoteScreen />
    </MemoryRouter>
  </Provider>
);

describe('Pruebas en <NoteScreen />', () => {

  test('debe de mostrarse correctamente', () => {

    expect(wrapper).toMatchSnapshot();


  });

  test('debe de disparar el active note', () => {

    wrapper.find('input[name="title"]').simulate("change", {
      target: {
        name: "title",
        value: "New title"
      }
    });

    expect(activeNote).toHaveBeenCalledWith('abc123', {
      body: 'Note body',
      title: 'New title',
      id: 'abc123'
    });

  });

});