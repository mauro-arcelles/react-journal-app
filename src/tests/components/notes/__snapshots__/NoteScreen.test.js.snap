// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      Thursday, April 14th 2022
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Note title"
    />
    <textarea
      autoComplete="off"
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happen today?"
      value="Note body"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
